### PR TITLE
[Risk mitigation] Rework components that display security results

### PR DESCRIPTION
**Pull request by schmanu**
## Overview
- redesigns balance change component
- components for rendering the redefine result
- new context that provides redesign scan result (and later adds other components)

## Screenshots
Screenshot 2023-05-26 at 13 09 02

### CODE REVIEW

File: src/components/tx/security/redefine/RedefineBalanceChange/index.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    return change.type === 'NATIVE'
+      ? item.tokenInfo.type === TokenType.NATIVE_TOKEN
+      : sameAddress(item.tokenInfo.address, change.address)
+  })?.tokenInfo.logoUri
+  return (
+    <>
+      <Typography>
+        {positive && '+'}
+        {formatVisualAmount(change.amount.value, change.decimals)}
+      </Typography>
```
Comment (iamacook) [inline src/components/tx/security/redefine/RedefineBalanceChange/index.tsx:36]: I'd format this as you do with the NFT balance change on lines 57-61

```suggestion
      <Typography mr={-1}>{positive ? '+' : '-'}</Typography>
      <Typography>
        {formatVisualAmount(change.amount.value, change.decimals)}
      </Typography>
```

File: src/components/tx/security/redefine/RedefineBalanceChange/index.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      {isLoading && !balanceChange ? (
+        <LoadingLabel />
+      ) : (
+        <Grid container direction="row" alignItems="center" spacing={1}>
+          {balanceChange ? (
+            <>
+              {balanceChange?.in?.map((change, idx) => (
+                <SingleBalanceChange change={change} key={idx} positive />
+              ))}
+              {balanceChange?.out?.map((change, idx) => (
```
Comment (iamacook) [inline src/components/tx/security/redefine/RedefineBalanceChange/index.tsx:132]:
```suggestion
              {balanceChange.in?.map((change, idx) => (
                <SingleBalanceChange change={change} key={idx} positive />
              ))}
              {balanceChange.out?.map((change, idx) => (
```

File: src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  const relevantWarnings = warnings.filter((warning) => warning.severity !== SecuritySeverity.NONE)
+  const shownWarnings =
+    relevantWarnings && relevantWarnings.length > MAX_SHOWN_WARNINGS
+      ? relevantWarnings.slice(0, MAX_SHOWN_WARNINGS)
+      : relevantWarnings
+  const hiddenWarningCount = warnings.length - shownWarnings.length
+  const hiddenMaxSeverity = hiddenWarningCount > 0 ? relevantWarnings[MAX_SHOWN_WARNINGS]?.severity : 0
+
+  const groupedShownWarnings = groupBy(shownWarnings, (warning) => warning.severity)
+  const sortedSeverities = Object.keys(groupedShownWarnings).sort((a, b) => (Number(a) < Number(b) ? 1 : -1))
```
Comment (iamacook) [inline src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx:39]: Perhaps we should memoise some of this. What do you think?

File: src/components/tx/security/redefine/useRedefine.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+export const useRedefine = (
+  safeTransaction: SafeTransaction | undefined,
+): AsyncResult<SecurityResponse<RedefineModuleResponse>> => {
+  const { safe, safeAddress } = useSafeInfo()
+  const wallet = useWallet()
+  const [retryCounter, setRetryCounter] = useState(0)
+  const isFeatureEnabled = useHasFeature(FEATURES.RISK_MITIGATION)
+
+  const redefineResponse = useAsync<SecurityResponse<RedefineModuleResponse>>(
```
Comment (iamacook) [inline src/components/tx/security/redefine/useRedefine.ts:33]: Nit: I'd suggest destructing the array here. It's hard to follow some of the code further down without doing so.

File: src/components/tx/security/shared/SecurityWarnings/index.tsx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    label: 'Medium issues',
+  },
+  [SecuritySeverity.LOW]: {
+    action: ACTION_REVIEW,
+    color: 'warning',
+    label: 'Low issues',
+  },
+  [SecuritySeverity.NONE]: {
+    color: 'info',
+    label: 'No issues found',
```
Comment (iamacook) [inline src/components/tx/security/shared/SecurityWarnings/index.tsx:57]: Nit: I'd pluralise this in the code if more than one warning is returned.

File: src/components/tx/security/TransactionSecurityContext/index.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}: {
+  children: JSX.Element
+  safeTx: SafeTransaction | undefined
+}) => {
+  const [redefineResponse, redefineError, redefineLoading] = useRedefine(safeTx)
+  const [isRiskConfirmed, setIsRiskConfirmed] = useState(false)
+
+  const providedValue = useMemo(
+    () => ({
+      verdict: redefineResponse?.severity,
```
Comment (iamacook) [inline src/components/tx/security/TransactionSecurityContext/index.tsx:41]: Nit: we use the `SecuritySeverity` enum so I'd name the property the same.

```suggestion
      severity: redefineResponse?.severity,
```

**Review by iamacook — COMMENTED** [review]
I'm sure you're aware but some tests are failing.

File: src/components/tx/AdvancedParams/index.tsx [hunk 7]
```diff
@@ -28,8 +28,7 @@ const AdvancedParams = ({
   willRelay,
 }: Props) => {
   const [isEditing, setIsEditing] = useState<boolean>(false)
-  const chain = useCurrentChain()
-  const isEIP1559 = !!chain && hasFeature(chain, FEATURES.EIP1559)
+  const isEIP1559 = !!useHasFeature(FEATURES.EIP1559)
```
Comment (usame-algan) [inline src/components/tx/AdvancedParams/index.tsx:31]: Why do we need `!!` here but not in the other places where we use this hook?

File: src/components/tx/security/TransactionSecurityContext/index.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+}>({
+  verdict: SecuritySeverity.NONE,
+  simulationUuid: undefined,
+  warnings: [],
+  balanceChange: undefined,
+  isLoading: false,
+  error: undefined,
+  needsRiskConfirmation: false,
+  isRiskConfirmed: false,
+  setIsRiskConfirmed: () => {},
```
Comment (usame-algan) [inline src/components/tx/security/TransactionSecurityContext/index.tsx:26]: nit: I would put this in the same order as the definitions above.

File: src/components/tx/security/redefine/RedefineBalanceChange/index.tsx [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+  change: NonNullable<RedefineModuleResponse['balanceChange']>['in' | 'out'][number] & { type: 'ERC20' | 'NATIVE' }
+  positive?: boolean
+}) => {
+  const { balances } = useBalances()
+
+  const logoUri = balances.items.find((item) => {
+    return change.type === 'NATIVE'
+      ? item.tokenInfo.type === TokenType.NATIVE_TOKEN
+      : sameAddress(item.tokenInfo.address, change.address)
+  })?.tokenInfo.logoUri
```
Comment (usame-algan) [inline src/components/tx/security/redefine/RedefineBalanceChange/index.tsx:30]:
```suggestion
  })?.tokenInfo.logoUri
  
```

File: src/components/tx/security/redefine/RedefineBalanceChange/index.tsx (hunk 2 above)
Comment (usame-algan) [reply src/components/tx/security/redefine/RedefineBalanceChange/index.tsx:132]: Could even remove the ? after `in` and `out` I think

File: src/components/tx/security/redefine/RedefineBalanceChange/index.tsx [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+      )}
+      <Typography variant="subtitle2" fontWeight={700} className={css.nftId}>
+        #{change.tokenId}
+      </Typography>
+      <Chip sx={{ borderRadius: '4px' }} label="NFT" />
+    </>
+  )
+}
+
+const SingleBalanceChange = ({
```
Comment (usame-algan) [inline src/components/tx/security/redefine/RedefineBalanceChange/index.tsx:82]: nit: If there is no `MultipleBalanceChanges` component or something similar you could also shorten the name to `BalanceChange` as its already singular.

File: src/components/tx/security/redefine/RedefineBalanceChange/index.tsx [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    <Box className={css.box}>
+      <Box className={css.head}>
+        <Grid container direction="row">
+          <Typography variant="subtitle2" fontWeight={700}>
+            Balance change
+          </Typography>
+        </Grid>
+      </Box>
+
+      {isLoading && !balanceChange ? (
```
Comment (usame-algan) [inline src/components/tx/security/redefine/RedefineBalanceChange/index.tsx:123]: If you extract this part of the render into its own component you could add early returns and simplify the nesting i.e.

```
if (isLoading) return <LoadingLabel />
if (balanceChange) ...
return ...
```

File: src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+import { SecuritySeverity } from '@/services/security/modules/types'
+import { groupBy } from 'lodash'
+import { Box } from '@mui/material'
+import ExternalLink from '@/components/common/ExternalLink'
+import { FEATURES } from '@/utils/chains'
+import { useHasFeature } from '@/hooks/useChains'
+
+const MAX_SHOWN_WARNINGS = 3
+
+const REDEFINE_SIMULATION_URL = 'https://app.redefine.net/defirewall-tx/'
```
Comment (usame-algan) [inline src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx:13]: I would suggest moving this into the const file

File: src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    needsRiskConfirmation,
+    isRiskConfirmed,
+    setIsRiskConfirmed,
+  } = useContext(TransactionSecurityContext)
+  const isFeatureEnabled = useHasFeature(FEATURES.RISK_MITIGATION)
+
+  /* Evaluate security warnings */
+  const relevantWarnings = warnings.filter((warning) => warning.severity !== SecuritySeverity.NONE)
+  const shownWarnings =
+    relevantWarnings && relevantWarnings.length > MAX_SHOWN_WARNINGS
```
Comment (usame-algan) [inline src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx:32]: According to the type `relevantWarnings` is always defined. Is it just an empty array in case there are no relevantWarnings? If so we can remove this check here.

Comment (usame-algan) [reply src/components/tx/security/redefine/RedefineScanResult/RedefineScanResult.tsx:32]: Also I am not sure if we need this condition at all. Can we not always just slice it? It shouldn't throw any errors even if there are fewer elements than what we slice off

File: src/hooks/useChains.ts [hunk 14]
```diff
@@ -25,3 +26,14 @@ export const useCurrentChain = (): ChainInfo | undefined => {
...
   return chainInfo
 }
+
+/**
+ * Checks if a feature is enabled on the current chain.
+ *
+ * @param feature name of the feature to check for
+ * @returns `true`, if the feature is enabled on the current chain. Otherwise `false`
+ */
+export const useHasFeature = (feature: FEATURES): boolean => {
```
Comment (usame-algan) [inline src/hooks/useChains.ts:36]: We still have a few places where `hasFeature` is used. Is this intended or should we update those to the hook as well?

File: public/images/transactions/redefine.svg
Comment (usame-algan) [inline public/images/transactions/redefine.svg:1]: Can we remove the white background on this logo? It is visible when hovering the accordion

**Review by usame-algan — COMMENTED** [review]
Looks great, well done! 👍 

I have a few suggestions

File: src/hooks/useChains.ts (hunk 14 above)
Comment (schmanu) [reply src/hooks/useChains.ts:36]: For now I updated all the components / hook which did not need the `chainInfo` at all but were using the pattern of 

```ts
const chainInfo = useCurrentChain()
const isEnabled = hasFeature(chainInfo, featureName)
```

some hooks actually need the chainInfo as well. Those ones I did not touch.
Sometimes `hasFeature` is also used outside of hooks / components.

File: public/images/transactions/redefine.svg
Comment (schmanu) [reply public/images/transactions/redefine.svg:1]: It will be replaced by a new logo shortly.

File: src/components/tx/security/shared/SecurityWarnings/styles.module.css [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+.verdictBox {
+  border: none;
+  border-radius: 6px;
+  border-bottom: 1px solid var(--color-border-light);
+}
+
+.verdictBox :global .MuiAccordionSummary-content {
+  margin-bottom: 0;
+}
```
Comment (usame-algan) [inline src/components/tx/security/shared/SecurityWarnings/styles.module.css:39]: This causes a small visual bug where its not centered anymore

Screenshot 2023-06-15 at 15 17 19

File: src/components/tx/security/shared/SecurityWarnings/index.tsx [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+              )
+            )}
+          </Grid>
+        </AccordionSummary>
+        <AccordionDetails>{children}</AccordionDetails>
+      </Accordion>
+      {needsConfirmation && (
+        <Box display="flex" flexDirection="row" gap={1} alignItems="center">
+          <Track {...MODALS_EVENTS.ACCEPT_RISK}>
+            <Checkbox checked={isConfirmed} onChange={() => setIsConfirmed(!isConfirmed)} />
```
Comment (usame-algan) [inline src/components/tx/security/shared/SecurityWarnings/index.tsx:156]: Can we wrap this in a `<FormControlLabel>` similar to the execute checkbox so the whole line is clickable? Otherwise the user has to click directly on the icon which is a bit harder.
